### PR TITLE
add ec3 clients

### PR DESCRIPTION
**Pull request by RytisSapka**
Added EC3 Clients, unit and integration tests.

### CODE REVIEW

File: src/reporting/clients/IEC3ConfigurationsClient.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import type { AccessToken } from "@itwin/core-bentley";
+import { EC3Configuration, EC3ConfigurationCreate, EC3ConfigurationUpdate } from "../interfaces/EC3Configurations";
+import type { EntityListIterator } from "../iterators/EntityListIterator";
+
+export interface IEC3ConfigurationsClient {
+  /**
+   * Gets all EC3 Configurations for an iModel. This method returns the full list of EC3 Configurations.
```
Comment (MindaugasButkus) [inline src/reporting/clients/IEC3ConfigurationsClient.ts:11]: EC3 configurations have nothing to do with iModels. They only refer to Reports. See https://developer.bentley.com/apis/carbon-calculation/operations/create-ec3-configuration/

File: src/test/MappingsClient.test.ts [hunk 2]
```diff
@@ -214,7 +214,7 @@ describe("mappings Client", () => {
 
     mapping = await mappingsClientNewBase.updateMapping("auth", "iModelId", "mappingId", newMapping);
     expect(fetchStub.calledWith(
-      "https://api.bentley.com/insights/reporting/datasources/imodels/iModelId/mappings/mappingId",
+      "BASE/datasources/imodels/iModelId/mappings/mappingId",
```
Comment (MindaugasButkus) [inline src/test/MappingsClient.test.ts:217]: What happened here?

File: src/test/EC3JobsClient.test.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import * as chaiAsPromised from "chai-as-promised";
+import { expect, use } from "chai";
+import * as sinon from "sinon";
+import { EC3JobsClient } from "../reporting/clients/EC3JobsClient";
+import { CarbonUploadState, EC3JobCreate } from "../reporting/interfaces/EC3Jobs";
+use(chaiAsPromised);
+
+describe("EC3 Jobs Client Client", () => {
```
Comment (MindaugasButkus) [inline src/test/EC3JobsClient.test.ts:12]: Duplicate `Client`. Also, make it match the class name (without spaces).

File: src/test/MappingsClient.test.ts (hunk 2 above)
Comment (MindaugasButkus) [reply src/test/MappingsClient.test.ts:217]: You are reusing the same `fetchStub` multiple times in this test, so you do not know which method called `fetchStub` with what parameters.

File: src/integration-test/integration/ReportsClient.test.ts [hunk 4]
```diff
@@ -105,14 +105,6 @@ describe("Reports Client", () => {
     expect(report.description).to.be.eq("Updated");
   });
 
-  it("Reports - Get all non deleted", async () => {
```
Comment (MindaugasButkus) [inline src/integration-test/integration/ReportsClient.test.ts:108]: Why did you remove this test?

File: src/test/EC3ConfigurationsClient.test.ts [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+describe("EC3 Configurations Client", () => {
+  const configurationsClient: EC3ConfigurationsClient = new EC3ConfigurationsClient();
+  const configurationsClientNewBase: EC3ConfigurationsClient = new EC3ConfigurationsClient("BASE");
+  let fetchStub: sinon.SinonStub;
+  let requestStub: sinon.SinonStub;
+
+  beforeEach(() => {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    fetchStub = sinon.stub(EC3ConfigurationsClient.prototype, "fetchJSON" as any);
```
Comment (MindaugasButkus) [inline src/test/EC3ConfigurationsClient.test.ts:20]: Have you tried stupping the `fetchJSON` method directly on the instance, not on the prototype?

File: src/test/MappingsClient.test.ts (hunk 2 above)
Comment (RytisSapka) [reply src/test/MappingsClient.test.ts:217]: I am reusing the fetch stub. I first fetch with the client that has the default base path ant then with the client that has ```BASE``` base path. When writing tests I noticed that this expect was not using the ```BASE``` path so I changed it (was epecting the same thing twice). Overall these tests should be remade using ```getcall(0).args``` because right now if they fail the message is ```Expected false be eqial to true```. Not sure if this should be done in this PR though.

File: src/integration-test/integration/ReportsClient.test.ts (hunk 4 above)
Comment (RytisSapka) [reply src/integration-test/integration/ReportsClient.test.ts:108]: Not sure how to correctly test this because if other tests create reports this test becomes polluted with them. Perhaps I should just check if the count is greater than 3?

File: src/test/MappingsClient.test.ts (hunk 2 above)
Comment (MindaugasButkus) [reply src/test/MappingsClient.test.ts:217]: What I wanted to point out that this change does not make sense. The test was not failing with the wrong expectation.

File: src/reporting/clients/IEC3ConfigurationsClient.ts (hunk 1 above)
Comment (RytisSapka) [reply src/reporting/clients/IEC3ConfigurationsClient.ts]: removed the ```for an iModel``` parts

File: src/integration-test/integration/ReportsClient.test.ts (hunk 4 above)
Comment (RytisSapka) [reply src/integration-test/integration/ReportsClient.test.ts:108]: Added test back.

File: src/test/MappingsClient.test.ts (hunk 2 above)
Comment (RytisSapka) [reply src/test/MappingsClient.test.ts]: Now changing base path is being handled in another test

File: src/test/EC3ConfigurationsClient.test.ts (hunk 5 above)
Comment (RytisSapka) [reply src/test/EC3ConfigurationsClient.test.ts]: stubbed on instance

File: src/reporting/clients/IEC3ConfigurationsClient.ts (hunk 1 above)
Comment (MindaugasButkus) [reply src/reporting/clients/IEC3ConfigurationsClient.ts]: Is this comment resolved?

image

Comment (RytisSapka) [reply src/reporting/clients/IEC3ConfigurationsClient.ts]: changed all iModel references to project references.

File: src/reporting/clients/IEC3ConfigurationsClient.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+   */
+  getConfigurations(
+    accessToken: AccessToken,
+    projectId: string,
+    top?: number
+  ): Promise<EC3Configuration[]>;
+
+  /**
+   * Gets an async paged iterator for EC3 Configurations within the context of a Project.
+   * This method returns an iterator which loads pages of mappings as it is being iterated over.
```
Comment (MindaugasButkus) [inline src/reporting/clients/IEC3ConfigurationsClient.ts:26]: The pages are most likely not `of mappings`.

File: src/reporting/clients/EC3ConfigurationsClient.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import { Collection, getEntityCollectionPage } from "../iterators/IteratorUtil";
+import { CARBON_CALCULATION_BASE_PATH, OperationsBase } from "../OperationsBase";
+import { IEC3ConfigurationsClient } from "./IEC3ConfigurationsClient";
+
+export class EC3ConfigurationsClient extends OperationsBase implements IEC3ConfigurationsClient {
+  constructor(basePath?: string) {
+    super(basePath ?? CARBON_CALCULATION_BASE_PATH);
+  }
+
+  public async getConfigurations(accessToken: string, iModelId: string, top?: number | undefined): Promise<EC3Configuration[]> {
```
Comment (Guillar1) [inline src/reporting/clients/EC3ConfigurationsClient.ts:18]: Still `iModelId` here.

File: src/reporting/clients/EC3ConfigurationsClient.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  public getConfigurationsIterator(accessToken: string, iModelId: string, top?: number | undefined): EntityListIterator<EC3Configuration> {
+    if(!this.topIsValid(top)) {
+      throw new RequiredError(
+        "top",
+        "Parameter top was outside of the valid range [1-1000]."
+      );
+    }
+    let url = `${this.basePath}/ec3/configurations?iTwinId=${iModelId}`;
```
Comment (Guillar1) [inline src/reporting/clients/EC3ConfigurationsClient.ts:34]: `iTwinId=${iModelId}` 🤔

File: src/reporting/clients/IEC3ConfigurationsClient.ts (hunk 6 above)
Comment (RytisSapka) [reply src/reporting/clients/IEC3ConfigurationsClient.ts]: Triple checked documentation

File: src/integration-test/integration/ReportsClient.test.ts [hunk 9]
```diff
@@ -108,9 +113,8 @@ describe("Reports Client", () => {
   it("Reports - Get all non deleted", async () => {
     const reports = await reportsClient.getReports(accessToken, projectId, undefined, true);
     expect(reports).to.not.be.undefined;
-    for(const report of reports) {
-      expect(["Test1", "Test2", "Test3", "Test"]).to.include(report.displayName);
-    }
+    expect(reports.length).to.be.gt(3);
+    expect(reports.some((x) => x.deleted)).to.be.true;
```
Comment (MindaugasButkus) [inline src/integration-test/integration/ReportsClient.test.ts:117]: So the test is "Get all non deleted" and you expect some reports to be deleted? Makes no sense.

File: src/test/EC3ConfigurationsClient.test.ts [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+use(chaiAsPromised);
+
+describe("EC3ConfigurationsClient", () => {
+  const configurationsClient: EC3ConfigurationsClient = new EC3ConfigurationsClient();
+  let fetchStub: sinon.SinonStub;
+  let requestStub: sinon.SinonStub;
+
+  beforeEach(() => {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    fetchStub = sinon.stub(configurationsClient, "fetchJSON" as any);
```
Comment (MindaugasButkus) [inline src/test/EC3ConfigurationsClient.test.ts:19]: Don't stub methods in the `beforeEach` if you are changing them in each test.

File: src/test/EC3ConfigurationsClient.test.ts [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+    requestStub = sinon.stub(client, "createRequest" as any);
+
+    const returns = {
+      configuration: {
+        id: 1,
+      },
+    };
+    fetchStub.resolves(returns);
+    await client.getConfiguration("auth", "configurationId");
+    expect(fetchStub.getCall(0).args[0]).to.match(new RegExp("^BASE"));
```
Comment (MindaugasButkus) [inline src/test/EC3ConfigurationsClient.test.ts:43]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith

I'd prefer not to use regex if it is possible.

File: src/test/EC3ConfigurationsClient.test.ts (hunk 10 above)
Comment (RytisSapka) [reply src/test/EC3ConfigurationsClient.test.ts:19]: I am only changing them in the single test that changes the base path of the client (because a new client is created ant therefore needs to be stubbed)

File: src/integration-test/integration/ReportsClient.test.ts (hunk 9 above)
Comment (RytisSapka) [reply src/integration-test/integration/ReportsClient.test.ts:117]: renamed test

File: src/test/EC3ConfigurationsClient.test.ts (hunk 11 above)
Comment (RytisSapka) [reply src/test/EC3ConfigurationsClient.test.ts]: comparing to substring

**Review by MindaugasButkus — APPROVED** [review]

**Review by MindaugasButkus — CHANGES_REQUESTED** [review]
Add new clients to `src/insights-client.ts` for them to be exported.

**Review by MindaugasButkus — CHANGES_REQUESTED** [review]
Move EC3 clients from under `reporting` directory. EC3 clients are not for reporting, but rather for CarbonCalculation.

**Comment by RytisSapka** [comment]
> Move EC3 clients from under `reporting` directory. EC3 clients are not for reporting, but rather for CarbonCalculation.

Reorganized. Moved shared code to common folder, renamed OneClickLCA folder to carbon-calculation and moved the new clients there.

File: src/common/OperationsBase.ts [hunk 12]
```diff
@@ -3,7 +3,7 @@
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
 import isomorphicFetch from "cross-fetch";
-import { DataType, ECProperty } from "./interfaces/GroupProperties";
+import { DataType, ECProperty } from "../reporting/interfaces/GroupProperties";
```
Comment (MindaugasButkus) [inline src/common/OperationsBase.ts:6]: Nothing under `common` should be importing anything from `reporting` or `carbon-calculation`.

Comment (RytisSapka) [reply src/common/OperationsBase.ts]: moved isValid function to mappingsClient since it was the only client using it.

File: CHANGELOG.md [hunk 13]
```diff
@@ -0,0 +1,9 @@
+# Change Log - @itwin/insights-client
+
+## 0.3.3
+Tue, 15 Nov 2022
+
+### Minor changes
```
Comment (MindaugasButkus) [inline CHANGELOG.md:6]:
```suggestion
### Patches
```

This is only a patch version bump, so I think it wouldn't be correct to call this `Minor changes`.

File: src/test/MappingsClient.test.ts [hunk 14]
```diff
@@ -793,7 +792,11 @@ describe("mappingsClient", () => {
   });
 
   it("isValid", () => {
-    const privateClient = new MappingsClient() as unknown as IOperationsBase;
+    interface MappingsBase {
+      isNullOrWhitespace(input: string | null | undefined): boolean;
+      isValidECProperty (prop: ECProperty): boolean;
```
Comment (MindaugasButkus) [inline src/test/MappingsClient.test.ts:797]: Inconsistent formatting.
